File: app/src/main/java/com/centaury/cataloguemovie/ui/favorite/fragment/FavoriteMovieFragment.java
package com.centaury.cataloguemovie.ui.favorite.fragment;


import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AlertDialog;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.centaury.cataloguemovie.R;
import com.centaury.cataloguemovie.data.local.entity.MovieEntity;
import com.centaury.cataloguemovie.di.Injectable;
import com.centaury.cataloguemovie.ui.favorite.adapter.FavoriteFragmentCallback;
import com.centaury.cataloguemovie.ui.favorite.adapter.FavoriteMovieAdapter;
import com.centaury.cataloguemovie.ui.favorite.viewmodel.FavoriteMovieViewModel;
import com.centaury.cataloguemovie.utils.Helper;
import com.facebook.shimmer.ShimmerFrameLayout;

import java.util.concurrent.ExecutionException;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/**
 * A simple {@link Fragment} subclass.
 */
public class FavoriteMovieFragment extends Fragment implements FavoriteFragmentCallback, Injectable {

    @BindView(R.id.rv_favmovie)
    RecyclerView mRvFavMovie;
    @BindView(R.id.shimmer_view_container)
    ShimmerFrameLayout mShimmerViewContainer;
    @BindView(R.id.empty_state)
    LinearLayout mEmptyState;
    private Unbinder unbinder;

    @Inject
    ViewModelProvider.Factory factory;

    private FavoriteMovieViewModel favoriteMovieViewModel;

    public FavoriteMovieFragment() {
        // Required empty public constructor
    }


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container,
                             Bundle savedInstanceState) {
        // Inflate the layout for this fragment
        View view = inflater.inflate(R.layout.fragment_favorite_movie, container, false);

        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onActivityCreated(@Nullable Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);
        if (getActivity() != null) {
            favoriteMovieViewModel = obtainViewModel(getActivity());

            FavoriteMovieAdapter favoriteMovieAdapter = new FavoriteMovieAdapter(getActivity(), this);
            favoriteMovieViewModel.getFavoriteMovie().observe(getActivity(), favoriteMovie -> {
                if (favoriteMovie != null) {
                    mShimmerViewContainer.stopShimmer();
                    mShimmerViewContainer.setVisibility(View.GONE);
                    toggleEmptyMovies(favoriteMovie.size());
                    favoriteMovieAdapter.submitList(favoriteMovie);
                    favoriteMovieAdapter.notifyDataSetChanged();
                }
            });

            mRvFavMovie.setLayoutManager(new LinearLayoutManager(getContext()));
            mRvFavMovie.setHasFixedSize(true);
            mRvFavMovie.setAdapter(favoriteMovieAdapter);
            mRvFavMovie.addItemDecoration(new Helper.TopItemDecoration(55));
        }
    }

    private void showDialogDeleteFavorite(int movieId) {
        MovieEntity movieEntity;
        try {
            movieEntity = favoriteMovieViewModel.getDetailFavMovie(movieId);

            LayoutInflater inflater = LayoutInflater.from(getContext());
            View view = inflater.inflate(R.layout.item_alert_dialog, null);

            AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
            builder.setView(view);

            TextView title = view.findViewById(R.id.alerttitle);
            title.setText(getString(R.string.txt_title_delete_dialog));

            builder.setCancelable(false)
                    .setPositiveButton(getString(R.string.btn_delete), (dialog, which) -> {
                        favoriteMovieViewModel.deleteFavoriteMovie(movieEntity);
                        dialog.dismiss();
                        Toast.makeText(getContext(), getString(R.string.txt_remove_movie), Toast.LENGTH_SHORT).show();
                    })
                    .setNegativeButton(getString(R.string.btn_cancel), (dialog, which) -> {
                        dialog.dismiss();
                    });
            AlertDialog alertDialog = builder.create();
            alertDialog.show();

        } catch (ExecutionException | InterruptedException e) {
            e.printStackTrace();
        }
    }

    @NonNull
    private FavoriteMovieViewModel obtainViewModel(FragmentActivity activity) {
        return ViewModelProviders.of(activity, factory).get(FavoriteMovieViewModel.class);
    }

    private void toggleEmptyMovies(int size) {
        if (size > 0) {
            mEmptyState.setVisibility(View.GONE);
            mRvFavMovie.setVisibility(View.VISIBLE);
        } else {
            mRvFavMovie.setVisibility(View.GONE);
            mEmptyState.setVisibility(View.VISIBLE);
        }
    }

    @Override
    public void onResume() {
        super.onResume();
        mShimmerViewContainer.startShimmer();
    }

    @Override
    public void onPause() {
        mShimmerViewContainer.stopShimmer();
        super.onPause();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onDeleteItemClick(int movieId) {
        if (getActivity() != null) {
            showDialogDeleteFavorite(movieId);
        }
    }
}

File: app/src/main/java/com/centaury/cataloguemovie/ui/favorite/fragment/FavoriteMovieFragment.kt
package com.centaury.cataloguemovie.ui.favorite.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.centaury.cataloguemovie.MovieCatalogueApp
import com.centaury.cataloguemovie.R
import com.centaury.cataloguemovie.di.component.DaggerFavoriteMovieComponent
import com.centaury.cataloguemovie.ui.favorite.adapter.FavoriteFragmentCallback
import com.centaury.cataloguemovie.ui.favorite.adapter.FavoriteMovieAdapter
import com.centaury.cataloguemovie.ui.favorite.viewmodel.FavoriteMovieViewModel
import com.centaury.cataloguemovie.utils.*
import com.centaury.domain.movies.model.MoviesEntity
import kotlinx.android.synthetic.main.fragment_favorite_movie.*
import kotlinx.android.synthetic.main.item_empty_state_placeholder.*
import javax.inject.Inject

/**
 * A simple [Fragment] subclass.
 */
class FavoriteMovieFragment : Fragment(), FavoriteFragmentCallback {

    @Inject
    lateinit var viewModelFactory: ViewModelProvider.Factory
    private lateinit var favoriteMovieViewModel: FavoriteMovieViewModel

    private var movieFavoriteData = arrayListOf<MoviesEntity>()
    private val favoriteMovieAdapter: FavoriteMovieAdapter by lazy {
        FavoriteMovieAdapter(movieFavoriteData, this)
    }

    private lateinit var titleDialog: TextView
    private var movie: MoviesEntity? = null

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        // Inflate the layout for this fragment
        return inflater.inflate(R.layout.fragment_favorite_movie, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initInjector()
        initView()
    }

    private fun initView() {
        favoriteMovieViewModel =
            ViewModelProvider(this, viewModelFactory)[FavoriteMovieViewModel::class.java]

        with(rv_fav_movie) {
            hasFixedSize()
            layoutManager = LinearLayoutManager(context)
            addItemDecoration(CommonUtils.TopItemDecoration(55))
            adapter = favoriteMovieAdapter
        }

        initObserver()

    }

    private fun initObserver() {
        favoriteMovieViewModel.state.observe(viewLifecycleOwner, { state ->
            when (state) {
                is LoaderState.ShowLoading -> {
                    shimmer_view_container.startShimmer()
                    shimmer_view_container.visible()
                }
                is LoaderState.HideLoading -> {
                    shimmer_view_container.stopShimmer()
                    shimmer_view_container.gone()
                }
            }
        })

        favoriteMovieViewModel.result.observe(viewLifecycleOwner, { result ->
            movieFavoriteData.clear()
            movieFavoriteData.addAll(result)
            CommonUtils.toggleEmptyState(result.size, empty_state, rv_fav_movie)
            favoriteMovieAdapter.notifyDataSetChanged()
        })

        favoriteMovieViewModel.error.observe(viewLifecycleOwner, { error ->
            context?.showToast(error)
        })

        favoriteMovieViewModel.resultMovieById.observe(viewLifecycleOwner, { resultMovieById ->
            movie = resultMovieById
        })

        favoriteMovieViewModel.errorMovieById.observe(viewLifecycleOwner, { errorMovieById ->
            context?.showToast(errorMovieById)
        })

        favoriteMovieViewModel.errorDeleteMovie.observe(viewLifecycleOwner, { errorDeleteMovie ->
            context?.showToast(errorDeleteMovie)
        })
    }

    private fun initInjector() {
        DaggerFavoriteMovieComponent.builder()
            .appComponent((activity?.application as MovieCatalogueApp).appComponent)
            .build()
            .inject(this)

    }

    private fun showDialogDeleteFavorite(movieId: Int) {
        favoriteMovieViewModel.getFavoriteMovieByIdContract(movieId)

        val customDialog: AlertDialog.Builder? = context?.let { AlertDialog.Builder(it) }
        val inflater = LayoutInflater.from(context)
        val view = inflater.inflate(R.layout.item_alert_dialog, null)

        titleDialog = view.findViewById(R.id.alert_title)
        titleDialog.text = context?.getString(R.string.txt_title_delete_dialog)

        customDialog?.apply {
            setView(view)
            setCancelable(false)
            setPositiveButton(R.string.btn_delete) { dialog, _ ->
                movie?.let { favoriteMovieViewModel.getDeleteFavoriteMovieContract(context, it) }
                dialog.dismiss()
                context.showToast(R.string.txt_movie_remove)
            }
            setNegativeButton(R.string.btn_cancel) { dialog, _ ->
                dialog.dismiss()
            }

        }
        val alertDialog = customDialog?.create()
        alertDialog?.show()
    }

    override fun onResume() {
        super.onResume()
        shimmer_view_container.startShimmer()
    }

    override fun onPause() {
        shimmer_view_container.stopShimmer()
        super.onPause()
    }

    override fun onDeleteItemClick(movieId: Int) {
        showDialogDeleteFavorite(movieId)
    }
}